---
title: "MSBT2 Cancer RNASeq Data Analysis"
author: "Kakembo Fredrick Elishama"
date: "9/10/2021"
output: pdf_document
---

# Loading the Packages
```{r loading packages, message=FALSE}
library(tidyverse)
library(DESeq2)
library(edgeR)
library(reshape2)
library(dendextend)
```


# Loading data

```{r import meta data, message=FALSE}
#setwd("~/Desktop/Academics/HackBio Fellowship/Transcriptomics/NeuroGenomics-HackBio/02-PracticalDiffExp/scripts/")
#setwd("~/Desktop/Kakembo/Cancer_RNASeq_DataAnalysis/")
setwd("~/Desktop/MSBT/Year2/Semester 1/Cancer Genetics and Genomics/Cancer_RNASeq_DataAnalysis/Cancer_RNASeq_DataAnalysis/")

list.files() #List files in the directory

#Reading the metadata file
meta <- read.csv("metadata.file1", sep="\t", row.names = 1, header = T, stringsAsFactors = T)
#meta <- read.csv("metadata.file", sep="\t")

levels(meta$cancer_type)

head(meta)
```


Next we want to read in the data. Each sample's counts are stored in a separate file. files into a count matrix

```{r import counts}
counts <- read.csv("hisat2/hisat2_counts.txt", sep="\t", header=T, row.names=1, comment.char = "#")
#head(counts)

#Check the column names
colnames(counts)

#Remove the unwanted columns
counts[c("Chr", "Start", "End", "Strand", "Length")] <- NULL

colnames(counts) <- gsub(".bam", "", colnames(counts))
colnames(counts) <- gsub("hisat2.", "", colnames(counts))

head(counts)

#Check if samples in the countdata matrix have a corresponding entry in the metadata file
all(rownames(meta) %in% colnames(counts))

#Check if the order of the samples in the matrix is similar to that in the metadata file.
all(colnames(counts) == rownames(meta))

#Since the order of the the counts file is not the same as that of the metadata file, 
# I rearrange the count data file to match the arrangement of the metadata file.
counts <- counts[, rownames(meta)]

#Check again
all(colnames(counts) == rownames(meta))


```



# Part 1: Breast Cancer vs Health Controls


```{r}
#Subset the medata to just breast cancer and Health controls
meta1 <- subset(meta,cancer_type=="Breast" | cancer_type=="HC" )

write.csv(meta1, "Metadata_BR_HC.file", sep = "\t")
meta1 <- read.csv("Metadata_BR_HC.file", row.names = 1, header = T, stringsAsFactors = T)

levels(meta1$cancer_type)
meta1$cancer_type <- relevel(meta1$cancer_type, ref = "HC")
meta1$cancer_type
#Subset the corresponing counts for the BC and HC
counts1 <- counts[, rownames(meta1)]

#Create the DESeq object
dds<-DESeqDataSetFromMatrix(countData = counts1,
                            colData = meta1,
                            design = ~ cancer_type ) 

```



# Differential Expression
We shall pass our DESeqDataset object through the **DESeq()** function which will first normalize our counts, then performs an estimation of size factors (which control for differences in the library size of the sequencing experiments) and finally  the estimation of dispersion for each gene by fitting the object to a generalized linear model,glm and producing the corresponding pvalues at a 90% confidence.    
Normalization of counts is such an import step because of 2 main reasons;

- Some genes are considerably larger than others hence we expect more reads to align onto these genes compared to smaller genes, something that is no where associated to the gene's expression profile.
- Some samples may be deeply sequenced as compared to others, hence we expect such samples to have more reads mapping to genes than others.

We then use the **`result()`** function to extract out the results from the DESeq() output.

```{r}
dds <- DESeq(dds)
#Extracting the results from the formed dds object
res <- results(dds)
head(res, 10)
```

## Understanding the results output in the res object.
The res object generated has 6 columns with the rownames being the individual genes. 

- **baseMean**; This is the avarage of the counts for a particular gene from the different samples.
- **log2FoldChange**; this is basically the log-ratio of a gene's or transcript's expression values in 2 different conditions. 
- **lfcSE**; this is just the standard error in determining the DE signifince level of a particular gene.
- **stat**; is the Wald Statistic
- **pvalue**; this is the raw p-value indicating how significant a gene is differentially expressed. 
- **padj**; this is the adjusted pvalue after applying the Benjamini-Hochberg (BH) adjustment for multiple comparisons. 

And for values assigned as NAs are based on the criteria below;

- If baseMean for all samples is zero, the p values and lfc will be assigned NA
- If a row contains any sample with extreme count outlier, the p-value and padj are set to NA
- If a row is filetered due to low counts then only the padj is set to NA


```{r}
#Generate a summary of res object
summary(res)
```

The summary of res above gives us an idea of the possible number of differentially expressed genes (both upregulated and downregulated), as well possible counts of outliers in the our data at 0.1 significance level by default. 


## Data Quaity assessment with plots
A quality assessment of our the results is one of the most crucial steps reccommended to be done before any further analysis is performed. This typically informs us how good or bad our data is and how best it is suited for the analysis to be performed and whether any results need to filtered out.

However with DESeq2, this is not so much of a major concern since the quality control is already implemented in the package function DESeq() that  caters for any low counts and outliers. 

Here we shall simply use plots to give us an idea of the data we are planning to use for our downstream analysis.    
The 2 common plots to achieve this include the heatmap and PCA.

For generating heatmaps and more quality plots it is reccommended to first transform our data. 

### Data Transformation
Testing for DE operates on raw counts, however for visualizations and clusterings, its better to work with
transformed count data. There are 2 common waays of transformation; which produce transformed data on
the log2 scale which has been normalized with respect to library size or other normalization factors.

- Variance Stabilizing Transformations (VST) and
- regularized logarithm, rlog

The goal of transformation is to _eliminate any variance arising when the mean is low_.

One common argument used during transformation is **blind** which tells whether transformation should be
blind to sample information specified in the the design (ie in an unbiased manner), hence using only the
intercept. However this is not the best choice especially if the counts difference are attributed to the design
and if one is to use the transformed data for downstream analysis. If `blind = FALSE` is set, we take into
account already estimated dispersion due to design as we are transforming the data, and make the process
much faster.

**In Comparison**: VST runs faster than rlog. If the library size of the samples and therefore their size factors
vary widely, the rlog transformation is a better option than VST. Both options produce log2 scale data which
has been normalized by the DESeq2 method with respect to library size.


```{r}
#Count data Transformations
vsd <- vst(dds, blind=FALSE)  #computing for Variance Stabilizing Transformation
rld <- rlog(dds, blind=FALSE) #Computing for regularised logarithm
#A look at the generated matrices in contrast to the original 
#Original count matrix
head(countdata)
#Count matrix for vsd
head(assay(vsd)) #Assay extracts a matrix of normalized counts
#Count matrix produced by rlog
head(assay(rld))
```

**Effects of transformation on the variance**
We will use the dispersion plot to explore the effects of transformation on the variance in our data. This plots the standard deviation of the transformed data across samples, against the mean, using the shifted logarithm transformation, the regularized log transformation and the variance stabilizing transformation.

**What we expect:** The shifted logarithm has elevated standard deviation in the lower count range, and the regularized log to a lesser extent, while for the variance stabilized data the standard deviation is roughly constant along the whole dynamic range.

```{r}
library(vsn)
#Dispersion Plots
#for the untransformed dds
ntd <- normTransform(dds)
meanSdPlot(assay(ntd))
#for vst
meanSdPlot(assay(vsd))
#For rlog
meanSdPlot(assay(rld))
```

We generally do not see much change between rlog and vst in this case, however there is a noticiable improvement in the the variance as compared to the untransformed data, shown by a fairly lower fitted line.

We will use the vst transformed data obtained from above for the plots, since its the most reccomended approach.

### 1. Heatmap of Count matrix.
We will first use the heatmap to explore the quality for the count matrix. 

```{r}
select <- order(rowMeans(counts(dds,normalized=TRUE)), decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds))
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
cluster_cols=FALSE, annotation_col=df)
```



## 2. Heatmap of Sample-Sample distances
This heatmap gives us an overview of the similarities and dissimilarities between the samples. 

```{r}
#Calculate sample-2-sample distances
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists) #Converting the dist object to matrix
rownames(sampleDistMatrix) <- vsd$Condition
colnames(sampleDistMatrix) <- rownames(meta) 
#Defining the colors
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
clustering_distance_rows=sampleDists,
clustering_distance_cols=sampleDists,
col=colors)
```

From this matrix we can generally observe a larger simarity between intra-sample groups eg more similarity between samples 40-41 and samples 37-39, and we can also note major dissimilarities between disease group compared to normal samples shown by much lighter colors. 

However we do note one unexpected difference here. Sample37 which belongs to the normal samples seems to be very different from the other samples both in the same category and different category as well, which is one thing we can't have a conclusive answer for yet as to wether the changes were generated after the experiment or they are phsyiological changes. 

## 3. Principal component plot of the samples
This plot is similar to the distane matrix above which will basically cluster the samples basing on their properties.   
This type of plot is useful for visualizing the overall effect of experimental covariates, relatedness and batch effects.
```{r}
plotPCA(vsd, intgroup="Condition")
```

From this plot we generally see disease samples clustering together, and 2 of the normal samples clustering together apart from one. 

We will redraw the plots to include their sample labels to be able to tell which sample is generally different among the normal. 

```{r}
#Including samples in the pca plot
#PCA Analysis
mycols <- brewer.pal(8, "Dark2")[1:length(unique(meta$Condition))]
rld_pca <- function (rld, intgroup = "cancer_type", ntop = 500, colors=NULL, legendpos="bottomleft", main="PCA Biplot", textcx=1, ...) {
require(RColorBrewer)
rv = rowVars(assay(rld))
select = order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
pca = prcomp(t(assay(rld)[select, ]))
fac = factor(apply(as.data.frame(colData(rld)[, intgroup, drop = FALSE]), 1, paste, collapse = " : "))
if (is.null(colors)) {
if (nlevels(fac) >= 3) {
colors = brewer.pal(nlevels(fac), "Paired")
} else {
colors = c("black", "red")
}
}
pc1var <- round(summary(pca)$importance[2,1]*100, digits=1)
pc2var <- round(summary(pca)$importance[2,2]*100, digits=1)
pc1lab <- paste0("PC1 (",as.character(pc1var),"%)")
pc2lab <- paste0("PC1 (",as.character(pc2var),"%)")
plot(PC2~PC1, data=as.data.frame(pca$x), bg=colors[fac], pch=21, xlab=pc1lab, ylab=pc2lab, main=main, ...)
with(as.data.frame(pca$x), textxy(PC1, PC2, labs=rownames(as.data.frame(pca$x)), cex=textcx))
legend(legendpos, legend=levels(fac), col=colors, pch=20)
}
rld_pca(rld, colors=mycols, intgroup="cancer_type", xlim=c(-75, 35))
```

After adding labels, we can clearly tell that indeed it was sample37 that was clustering differently compared to the rest. 

## 4. Dispersion plot
This plot generally show how gene expression data differs across samples in the same treatment group.

- The black dots are the dispersion estimates for each gene separately
- The red line is the fitted trend which shows the dispersion’s dependence on the mean.
- The blue dots are genes that have been shrunk/fitted towards the red line.
- The outer blue circles with black dots inside are considered outliers and therefore not shrunk towards the fitted line

```{r}
#Dispersion plot
plotDispEsts(dds, main="Dispersion plot for DESeq")
```

## Exploring the results with plots

## MA plots
This plot is a quick visual for genes that are upregulated as well those downregulated, possible outliers, and those condidered not significant. Its a better practice to plot the shrinked values for this this plot as compared to the initial results. 

### Log fold change shrinkage
To generate plots and ranking of genes, its better we first perform a Shrinkage of effect size (ie the LFC estimates) that helps in shrinking any low counts/ low effect size that are close to zero towards the zero line. This helps in cleaning out the plots (very useful for the MA plot) for easier interpretation *without even the need of filtering out low counts in the results*.

> There are generally 3 methods for performing effect size shrinkage ie using the **`apeglm`** which is usally the default method, **`ashr`** and **`normal`** which was the initial method implemented by DESeq2. Among this the best recommended is the apeglm and most discouraged for use is normal.
For exploration purposes we will test all of the above methods on our results to better understand the difference it makes (if any).

One thing we need to provide to the `lfcshrink()` function on top of the method to use is the coefficient obtained from the `resultsNames()` which basically tells the shrink function what Condition is to be compared against what. 

```{r}
#Obtain the coefficient to use for shrinkage
resultsNames(dds)
```

In our case its `Condition_disease_vs_normal` or the second object in the output which basically tells us that we are comparing diseased samples against normal samples.

```{r}
#Computing for size shrikage using apeglm
resLFC <- lfcShrink(dds, coef="Condition_disease_vs_normal", type="apeglm")
head(resLFC)
```

Lets now generate the MA plot for the shrinked value.

```{r}
#MA plot for shrinked values
plotMA(resLFC, ylim=c(-2,2))
```

This plot basically has 3 features of our interest explained below. Each dot represented here is a gene.

1. The `grey` dots represent the genes whose LFC is considered not signicant to our condition.
2. The `blue` dots represent genes whose LFC is considered significant. Dots above the line represent upregulated genes, and dots below the line represent downregulated genes
3. The genes represented with triangles either coloured or not are considered as outliers basing on the limit we have set (in this case its from -2 to 2).


We shall now visualize the MA plot with first the initial res object where we hadn't performed the size shrinkage and compare it to when we have performed shrinkage using apeglm.
```{r}
# comparison of MA plot for unshrinked values and shrinked valued
par(mfrow=c(1,2))
plotMA(res, ylim=c(-2,2), main="Unshrunken")  #Unshrinked values
plotMA(resLFC, ylim=c(-2,2), main="Shrunken ")  #Shrinked values
```

One quick observation we can give from the plots above is that we have alot of unsignificants genes being represented in the unshrunken plot as compared to the shrunken. This is the exact reason why we perform size shrinkage, ie clear the plots and make it more easily interpretable. 

**Compare the shrinkage methods**    
We will now compute the size shrinkage using the other 2 methods and determine the difference it would make. 
```{r}
#Alternative shrinkage estimators
#using the ashr method
resAsh <- lfcShrink(dds, coef=2, type="ashr")
#Using the normal method
resNorm <- lfcShrink(dds, coef=2, type="normal")
#Generating MA plots comparing the 3 shrinkage estimators
#mar sets the margin sizes in the following order: bottom, left, top, right
par(mfrow=c(1,3), mar=c(4,4,2,1)) 
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resLFC, xlim=xlim, ylim=ylim, main="apeglm")
plotMA(resNorm, xlim=xlim, ylim=ylim, main="normal")
plotMA(resAsh, xlim=xlim, ylim=ylim, main="ashr")
```

Among these plots, the plot for shrinakge using the normal estimate contains more unsignificant genes followed by ashr
which would probably explain why apeglm is a better prefered estimator. 


## Exploring Significant Results

Lets first make a histogram plot of p-value frequency. 
```{r}
#Exploring p-value frequency
hist(res$pvalue, breaks=50, col="grey")
```

We can see from this that we have some good number of genes whose pvalue is significant (ie fall below 0.1), however most of the pvalues for the genes is not considered significant. 


From our res object we will go head to only subset out the significant results. We will consider a gene's expression significant if its pvalue is below 0.05, ie 95% confidence.

```{r}
#Reorder res based on pvalues
resOrdered <- res[order(res$pvalue),]
head(resOrdered)
#Only subsettting genes with Significant padj value
resSig <- subset(resOrdered, padj < 0.05)
#Sum of significant genes
nrow(resSig)
```

## Plot Counts
This plots the normalized counts for each sample in the particular gene

Lets start by plotting the counts of the most significant gene ie the one with the lowest pvalue.


```{r}
#For the gene with the minimum padj
plotCounts(dds, gene=which.min(res$padj), intgroup="cancer_type")
```

From this we can see that normal samples are having a fairly equal count for the gene as well as the disease samples, however there is a clear cut difference between the counts for this gene in the 2 categories. 


```{r}
View(subset(as.data.frame(resSig),log2FoldChange>=2))

BC_HC_sig <- subset(as.data.frame(resSig),log2FoldChange>=2)

write.csv(BC_HC_sig, "BC_vs_HC_significant.csv")
```





# Part 2: Lung Cancer vs Health Controls


```{r}
#Subset the medata to just breast cancer and Health controls
meta1 <- subset(meta,cancer_type=="Lung" | cancer_type=="HC" )

write.csv(meta1, "Metadata_Lung_HC.file", sep = "\t")
meta1 <- read.csv("Metadata_Lung_HC.file", row.names = 1, header = T, stringsAsFactors = T)

levels(meta1$cancer_type)
meta1$cancer_type <- relevel(meta1$cancer_type, ref = "HC")
meta1$cancer_type
#Subset the corresponing counts for the BC and HC
counts1 <- counts[, rownames(meta1)]

#Create the DESeq object
dds1<-DESeqDataSetFromMatrix(countData = counts1,
                            colData = meta1,
                            design = ~ cancer_type ) 

dds1<-DESeq(dds1)
```

```{r}
#Sample Dendogram
vsd <- vst(dds1, blind =FALSE)
mat <-assay(vsd)
head(as.data.frame(mat))


dend = t(mat) %>% 
  scale %>% 
  dist %>% 
  hclust %>% 
  as.dendrogram 

dend %>% 
  set("labels", meta1$cancer_type) %>% 
  plot
```


```{r}
mat.pca<-prcomp(t(mat)) #perform PCA
summary(mat.pca)

#Extract PCA scrores
scores <-as.data.frame(mat.pca$x) 
scores


# scores %>% 
#   mutate(GEO_Accession = rownames(scores)) %>% 
#   left_join(meta1) %>% 
#   ggplot(aes(PC1, PC2, color = cancer_type))+
#   geom_point()

```



```{r}
res <- results(dds1)
summary(res)


vsd <- vst(dds1, blind=FALSE)  #computing for Variance Stabilizing Transformation
rld <- rlog(dds1, blind=FALSE) #Computing for regularised logarithm

#BiocManager::install("vsn")
#BiocManager::install("pheatmap")
#BiocManager::install("calibrate")
library(vsn)
library(pheatmap)
library("RColorBrewer")
require(genefilter)
require(calibrate)
#Dispersion Plots
#for the untransformed dds
ntd <- normTransform(dds1)
meanSdPlot(assay(ntd))
#for vst
meanSdPlot(assay(vsd))
#For rlog
meanSdPlot(assay(rld))


select <- order(rowMeans(counts(dds1,normalized=TRUE)), decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds1))
#pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,cluster_cols=FALSE, annotation_col=df)


#Calculate sample-2-sample distances
sampleDists <- dist(t(assay(vsd)))


sampleDistMatrix <- as.matrix(sampleDists) #Converting the dist object to matrix
rownames(sampleDistMatrix) <- vsd$Condition
colnames(sampleDistMatrix) <- rownames(meta1) 
#Defining the colors
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sampleDistMatrix, clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, col=colors)


```


```{r}
#plotPCA(vsd, intgroup="Condition")

mycols <- brewer.pal(8, "Dark2")[1:length(unique(meta$Condition))]

rld_pca <- function (rld, intgroup = "Condition", ntop = 500, colors=NULL, legendpos="bottomleft", main="PCA Biplot", textcx=1, ...) {

require(RColorBrewer)
rv = rowVars(assay(rld))
select = order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
pca = prcomp(t(assay(rld)[select, ]))
fac = factor(apply(as.data.frame(colData(rld)[, intgroup, drop = FALSE]), 1, paste, collapse = " : "))
if (is.null(colors)) {
if (nlevels(fac) >= 3) {
colors = brewer.pal(nlevels(fac), "Paired")
} else {
colors = c("black", "red")
}
}
pc1var <- round(summary(pca)$importance[2,1]*100, digits=1)
pc2var <- round(summary(pca)$importance[2,2]*100, digits=1)
pc1lab <- paste0("PC1 (",as.character(pc1var),"%)")
pc2lab <- paste0("PC1 (",as.character(pc2var),"%)")
plot(PC2~PC1, data=as.data.frame(pca$x), bg=colors[fac], pch=21, xlab=pc1lab, ylab=pc2lab, main=main, ...)
with(as.data.frame(pca$x), textxy(PC1, PC2, labs=rownames(as.data.frame(pca$x)), cex=textcx))
legend(legendpos, legend=levels(fac), col=colors, pch=20)
}
rld_pca(rld, colors=mycols, intgroup="cancer_type", xlim=c(-75, 35))

```

```{r}
#Dispersion plot
plotDispEsts(dds1, main="Dispersion plot for DESeq")
```

```{r}
#Obtain the coefficient to use for shrinkage
resultsNames(dds1)


#Computing for size shrikage using apeglm
resLFC <- lfcShrink(dds1, coef="cancer_type_Lung_vs_HC", type="apeglm")
head(resLFC)


#MA plot for shrinked values
plotMA(resLFC, ylim=c(-1,2))
```


We shall now visualize the MA plot with first the initial res object where we hadn't performed the size shrinkage and compare it to when we have performed shrinkage using apeglm.
```{r}
# comparison of MA plot for unshrinked values and shrinked valued
par(mfrow=c(1,2))
plotMA(res, ylim=c(-2,2), main="Unshrunken")  #Unshrinked values
plotMA(resLFC, ylim=c(-2,2), main="Shrunken ")  #Shrinked values
```


## Exploring Significant Results

Lets first make a histogram plot of p-value frequency. 
```{r}
#Exploring p-value frequency
hist(res$pvalue, breaks=50, col="grey")

resOrdered <- res[order(res$pvalue),]
head(resOrdered)
#Only subsettting genes with Significant padj value
resSig <- subset(resOrdered, padj < 0.05)
#Sum of significant genes
nrow(resSig)

```


```{r}
#For the gene with the minimum padj
plotCounts(dds1, gene=which.min(res$padj), intgroup="cancer_type")
```


```{r}
View(subset(as.data.frame(resSig),log2FoldChange>=2))
dim(subset(as.data.frame(resSig),log2FoldChange>=2))
Lung_HC_sig <- subset(as.data.frame(resSig),log2FoldChange>=2)

write.csv(Lung_HC_sig, "Lung_vs_HC_significant.csv")
```
